src/ucp/proto/proto.h

```diff
@@ -0,0 +1,184 @@
+                                             function of message size in bytes */
+} ucp_proto_perf_range_t;
+
+
+/**
+ * UCP protocol capabilities (per operation parameters)
+ */
+typedef struct {
+    size_t                  cfg_thresh; /* Configured protocol threshold */
+    size_t                  min_length; /* Minimal message size */
+    unsigned                num_ranges; /* Number of entries in 'ranges' */
+
+    /* Performance estimation function for different message sizes */
+    ucp_proto_perf_range_t  ranges[UCP_PROTO_MAX_PERF_RANGES];
+
+} ucp_proto_caps_t;
+
+
+/**
+ * Parameters for protocol initialization function
+ */
+typedef struct {
+    /* Input parameters */
+    ucp_worker_h                   worker;         /* Worker to initialize on */
+    const ucp_proto_select_param_t *sel_param;     /* Operation parameters */
+    const ucp_ep_config_key_t      *ep_config_key; /* Endpoint configuration */
+
+    /* Output parameters */
+    void                           *priv;       /* Pointer to priv buffer */
+    size_t                         *priv_size;  /* Occupied size in priv buffer */
+    ucp_proto_caps_t               *caps;       /* Protocol capabilities */
+} ucp_proto_init_params_t;
+
+
+/**
+ * Initialize protocol-specific configuration and estimate protocol performance
+ * as function of message size.
+ *
+ * @param [in]  params   Protocol initialization parameters.
+ *
+ * @return UCS_OK              - if successful.
+ *         UCS_ERR_UNSUPPORTED - if the protocol is not supported on the key.
+ */
+typedef ucs_status_t
+(*ucp_proto_init_func_t)(const ucp_proto_init_params_t *params);
+
+
+/**
+ * Dump protocol-specific configuration.
+ *
+ * @param [in]  priv      Protocol private data, which was previously filled by
+ *                        @ref ucp_proto_init_func_t.
+ * @param [out] strb      Filled with a string of protocol configuration text.
+ *                        The user is responsible to release the string by
+ *                        calling @ref ucs_string_buffer_cleanup.
+ */
+typedef void
+(*ucp_proto_config_str_func_t)(const void *priv, ucs_string_buffer_t *strb);
+
+
+/**
+ * UCP base protocol definition
+ */
+struct ucp_proto {
+    const char                      *name;      /* Protocol name */
+    ucp_proto_init_func_t           init;       /* Initialization function */
+    ucp_proto_config_str_func_t     config_str; /* Configuration dump function */
+    uct_pending_callback_t          progress;   /* UCT progress function */
+};
+
+
+/**
+ * Register a protocol definition
+ */
+#define UCP_PROTO_REGISTER(_proto) \
+    UCS_STATIC_INIT { \
+        ucs_assert_always(ucp_protocols_count < UCP_PROTO_MAX_COUNT); \
+        ucp_protocols[ucp_protocols_count++] = (_proto); \
+    }
+
+
+/**
+ * Retrieve a protocol field by protocol id.
+ */
+#define ucp_proto_id_field(_proto_id, _field) \
+    (ucp_protocols[(_proto_id)]->_field)
+
+
+/**
+ * Call a protocol method by protocol id.
```

Comment:
don't need the trailing `.` as in other comments

Reply:
actually, we need comma for function (or macro) description since it describe an "action"

Reply:
ic, thank you for the explanation